exp/controllers/awsmachinepool_controller.go

```diff
@@ -415,6 +448,164 @@ func (r *AWSMachinePoolReconciler) reconcileDelete(machinePoolScope *scope.Machi
 	return nil
 }
 
+func reconcileDeleteAWSMachines(ctx context.Context, mp *expclusterv1.MachinePool, clusterScope cloud.ClusterScoper, client client.Client) error {
+	awsMachineList, err := getAWSMachines(ctx, mp, client)
+	if err != nil {
+		return err
+	}
+	for i := range awsMachineList.Items {
+		awsMachine := awsMachineList.Items[i]
+		if !awsMachine.DeletionTimestamp.IsZero() {
+			// delete the owner Machine resource for the AWSMachine so that CAPI can clean up gracefully
+			machine, err := util.GetOwnerMachine(ctx, client, awsMachine.ObjectMeta)
+			if err != nil {
+				clusterScope.Warn("failed to get owner machine", "awsmachine", awsMachine, "machinepool", mp.Name, "namespace", mp.Namespace)
+				continue
+			}
+
+			if err := client.Delete(ctx, machine); err != nil {
+				clusterScope.Warn("failed to delete owner machine", "awsmachine", awsMachine, "machinepool", mp.Name, "namespace", mp.Namespace)
+			}
+		}
+	}
+	return nil
+}
+
+func getAWSMachines(ctx context.Context, mp *expclusterv1.MachinePool, kubeClient client.Client) (*infrav1.AWSMachineList, error) {
+	awsMachineList := &infrav1.AWSMachineList{}
+	labels := map[string]string{
+		clusterv1.MachinePoolNameLabel: mp.Name,
+		clusterv1.ClusterNameLabel:     mp.Spec.ClusterName,
+	}
+	if err := kubeClient.List(ctx, awsMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+	return awsMachineList, nil
+}
+
+func createAWSMachinesIfNotExists(ctx context.Context, awsMachineList *infrav1.AWSMachineList, mp *expclusterv1.MachinePool, existingASG *expinfrav1.AutoScalingGroup, clusterScope cloud.ClusterScoper, client client.Client, ec2Svc services.EC2Interface) error {
+	clusterScope.Trace("creating missing awsmachines")
+
+	providerIDs := make(map[string]infrav1.AWSMachine, len(awsMachineList.Items))
+	for i := range awsMachineList.Items {
+		awsMachine := awsMachineList.Items[i]
+		if awsMachine.Spec.ProviderID == nil || *awsMachine.Spec.ProviderID == "" {
+			continue
+		}
+		providerID := *awsMachine.Spec.ProviderID
+		providerIDs[providerID] = awsMachine
+	}
+	ownerRef := metav1.OwnerReference{
+		APIVersion: mp.GroupVersionKind().Version,
+		Kind:       mp.Kind,
+		Name:       mp.Name,
+		UID:        mp.UID,
+	}
+
+	for i := range existingASG.Instances {
+		instanceID := existingASG.Instances[i].ID
+		providerID := fmt.Sprintf("aws:///%s/%s", existingASG.Instances[i].AvailabilityZone, instanceID)
+
+		clusterScope.Trace("checking if machinepool awsmachine is up to date", "providerID", providerID, "instanceID", instanceID, "asg", existingASG.Name)
+		if _, exists := providerIDs[providerID]; exists {
+			continue
+		}
+
+		instance, err := ec2Svc.InstanceIfExists(&instanceID)
+		if err != nil {
```

Comment:
I added a guard clause to try to handle this:
```
		if errors.Is(err, ec2.ErrInstanceNotFoundByID) {
			instanceLogger.V(5).Info("instance not found, it may have already been deleted.")
			continue
		}
```